File: Hard/Javascript/135. Candy.js
/**There are n children standing in a line. Each child is assigned a rating value given in the integer array ratings.

You are giving candies to these children subjected to the following requirements:

Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.
Return the minimum number of candies you need to have to distribute the candies to the children. */

/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function (ratings) {
  let candy = Array(ratings.length).fill(1);

  for (let i = 1; i < ratings.length; i++) {
    if (ratings[i] > ratings[i - 1]) {
      candy[i] = candy[i - 1] + 1;
    }
  }

  for (let i = ratings.length - 2; i > -1; i--) {
    if (ratings[i] > ratings[i + 1] && candy[i] <= candy[i + 1]) {
      candy[i] = candy[i + 1] + 1;
    }
  }

  return candy.reduce((total, val) => total + val,0);
};

console.log(candy([1, 0, 2]));

File: Hard/Python/135. Candy.py
'''
There are n children standing in a line. Each child is assigned a rating value given in the integer array ratings.

You are giving candies to these children subjected to the following requirements:

Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.
Return the minimum number of candies you need to have to distribute the candies to the children.
'''

class Solution:
    def candy(self, ratings: list[int]) -> int:
        candy = [1] * len(ratings)

        for ind in range(1, len(ratings)):
            if ratings[ind] > ratings[ind -1]:
                candy[ind] = candy[ind-1] + 1

        for ind in range(len(ratings) - 2, -1, -1):
            if ratings[ind] > ratings[ind + 1] and candy[ind] <= candy[ind + 1]:
                candy[ind] = candy[ind+1] + 1

        return sum(candy)

print(Solution().candy([10,9,8,7,6,5,4,3,2,1]))